packages/financial-templates-lib/src/price-feed/BigQueryPriceFeed.js

```diff
@@ -0,0 +1,219 @@
+
+    return returnPrice;
+  }
+
+  getLastUpdateTime() {
+    return this.lastUpdateTime;
+  }
+
+  async update() {
+    // Using current time minus 5 minutes as end-bound of query. This is because the BQ dataset lags ~5 minutes.
+    // Using moment package to change to UTC and create acceptable format for BQ.
+    const currentTime = new Date();
+    const formattedCurrentTime = moment(currentTime)
+      .subtract(5, "minutes")
+      .utc()
+      .format("YYYY-MM-DD HH:mm:ss");
+
+    // Subtracting 30 days from current time to create the earlier time bound.
+    let t2 = new Date();
+    t2 = moment(t2)
+      .subtract(30, "days")
+      .utc()
+      .format("YYYY-MM-DD HH:mm:ss");
+
+    // Return early if the last call was too recent.
+    if (this.lastUpdateTime !== undefined && this.lastUpdateTime + this.minTimeBetweenUpdates > currentTime) {
+      this.logger.debug({
+        at: "BigQueryPriceFeed",
+        message: "Update skipped because the last one was too recent",
+        currentTime: currentTime,
+        lastUpdateTimestamp: this.lastUpdateTime,
+        timeRemainingUntilUpdate: this.lastUpdateTimes + this.minTimeBetweenUpdates - currentTime
+      });
+      return;
+    }
+
+    this.logger.debug({
+      at: "BigQueryPriceFeed",
+      message: "Updating",
+      currentTime: currentTime,
+      lastUpdateTimestamp: this.lastUpdateTime
+    });
+
+    // Create the query with the needed time interval.
+    const query = this.createQuery(t2, formattedCurrentTime);
+
+    // Submit async call to BigQuery.
+    let priceResponse;
+    await this.runQuery(query)
+      .then(res => {
+        console.log(res[0].gas_price);
+        priceResponse = res[0].gas_price;
+      })
+      .catch(console.log);
+
+    // Check responses.
+    if (!priceResponse) {
+      throw new Error(`🚨Could not parse price result from bigquery: ${priceResponse}`);
+    }
+
+    const newPrice = this.convertDecimals(priceResponse);
+
+    // Store results.
+    this.currentPrice = newPrice;
+    this.lastUpdateTime = currentTime;
+  }
+  async runQuery(query) {
+    // returns a node read stream
+    const stream = await client.createQueryStream({ query });
+    // highland wraps a stream and adds utilities simlar to lodash
+    // https://caolan.github.io/highland/
+    return (
+      highland(stream)
+        // from here you can map or reduce or whatever you need for down stream processing
+        // we are just going to "collect" stream into an array for display
+        .collect()
+        // emit the stream as a promise when the stream ends
+        // this is the start of a data pipeline so you can imagine
+        // this could also "pipe" into some other processing pipeline or write to a file
+        .toPromise(Promise)
+    );
+  }
+  // This is a helper method to create a GASETH BQ query with time arguments.
+  createQuery(t2, formattedCurrentTime) {
+    const query = `
```

Comment:
Should we place queries in an external Query.js file? This seems like something we might want to tweak in the future so it might be easier to find in another file

Reply:
+1

Reply:
Created a separate file in a new folder `../queries/GasEthQuery.js`.